File: pages/translator_page.py
# translator_page.py
import logging

logger = logging.getLogger(__name__)
import time
from enum import Enum, auto
from typing import ClassVar

from PySide6.QtCore import (
    QObject,
    Qt,
    QThread,
    Signal,
    Slot,
)
from PySide6.QtWidgets import QHBoxLayout

from core.hotkey_manager import hotkey_manager
from core.page_router import page_router
from pages.base_page import BasePage
from resources.constants import CONFIG
from resources.svgs import arrow_right_icon, translate_icon
from utils import text_manager
from utils.translator import Translator
from widgets.core_button import CoreButton
from widgets.selection_grid import SelectionGrid
from widgets.svg_button import SvgButton
from widgets.text_editor import RoundedTextEdit


class _GlobalSignals(QObject):
    translate_shortcut_signal = Signal(str)


# 模块级单例
translation_global_signals = _GlobalSignals()


class TranslationHotkey(QObject):
    """一键翻译全局热键

    按下快捷键后自动完成：复制选中文本 → 填入输入框 → 使用默认服务翻译。
    pynput 的回调运行在监听线程，不能直接操作 Qt 控件，
    因此通过信号以 QueuedConnection 转发到 Qt 主线程执行。
    """

    _triggered = Signal()

    def __init__(self, callback, hotkey: str | None = None, parent=None):
        super().__init__(parent)
        self._callback = callback
        self._hotkey = hotkey or CONFIG["translator"].get("hotkey", "ctrl+shift+t")
        self._registered = False
        self._triggered.connect(self._run_in_main_thread, Qt.ConnectionType.QueuedConnection)

    @property
    def hotkey(self) -> str:
        return self._hotkey

    def start(self):
        """注册全局热键并启动全局键盘监听（幂等）"""
        if hotkey_manager.register(self._hotkey, self._fire):
            self._registered = True
            logger.info(f"[TranslationHotkey] 一键翻译已启用，快捷键: {self._hotkey}")
        else:
            logger.error(f"[TranslationHotkey] 一键翻译快捷键 {self._hotkey} 注册失败！")

    def stop(self):
        """注销全局热键"""
        if self._registered:
            hotkey_manager.unregister(self._hotkey)
            self._registered = False

    def _fire(self):
        """pynput 监听线程回调：仅转发信号，不做任何 Qt 操作"""
        self._triggered.emit()

    @Slot()
    def _run_in_main_thread(self):
        """Qt 主线程中执行实际的一键翻译流程"""
        if self._callback:
            self._callback()


class TranslationWorker(QThread):
    """后台翻译线程

    translate_text 是同步阻塞调用（含网络请求），直接在主线程执行会卡死 UI。
    放入 QThread 执行，完成后通过信号把结果回传主线程；
    调用 cancel() 后结果将被丢弃，不再更新界面。
    """

    translation_done = Signal(object)

    def __init__(self, translator, text, server, from_lang, to_lang, parent=None):
        super().__init__(parent)
        self._translator = translator
        self._text = text
        self._server = server
        self._from_lang = from_lang
        self._to_lang = to_lang
        self._cancelled = False

    def cancel(self):
        """请求取消：置标志，翻译结果将被丢弃（同步请求无法中断网络传输）"""
        self._cancelled = True

    def run(self):
        if self._cancelled:
            return
        result = self._translator.translate_text(
            text=self._text,
            server=self._server,
            from_lang=self._from_lang,
            to_lang=self._to_lang,
        )
        if not self._cancelled:
            self.translation_done.emit(result)


class TranslatorPage(BasePage):
    PAGE_NAME = "translator"
    TITLE = "Translator"
    MODULE_NAME = "Translator"
    MODULE_ICON = translate_icon

    SUPPORTED_LANGUAGES: ClassVar[list[str]] = [
        "Auto",
        "English",
        "Chinese",
        "Japanese",
        "Korean",
        "French",
        "German",
        "Spanish",
        "Russian",
    ]
    SUPPORTED_SERVERS: ClassVar[list[str]] = ["Google", "DeepL", "Baidu", "Bing", "AI1", "AI2"]

    RESULT_TEXT_HEIGHT = 120

    def __init__(self, parent=None):
        super().__init__(parent)
        self.translator = Translator()
        self.target_size = (400, 300)

        self._worker = None
        self._translation_cancelled = False
        self._server_display_to_id: dict[str, str] = {}

        layout = self.set_main_layout("v")
        assert layout is not None

        self.input_text = RoundedTextEdit(placeholder="Enter or paste text here...", parent=self)
        self.result_text = RoundedTextEdit(placeholder="Translation result", parent=self)
        self.result_text.setFixedHeight(0)

        self.selection_grid = SelectionGrid()
        # 点击网格项后，直接调用 grid 本身的收起逻辑
        self.selection_grid.item_selected.connect(lambda _: self.selection_grid.collapse())

        self.footer_layout = QHBoxLayout()
        self.footer_layout.addStretch()

        # 原语种按钮：启用原生 Checkable
        self.origin_lang = CoreButton(text=CONFIG["translator"]["default_from_lang"])
        self.origin_lang.setCheckable(True)
        self.origin_lang.clicked.connect(lambda: self.display_lang_list("origin"))
        self.footer_layout.addWidget(self.origin_lang, alignment=Qt.AlignmentFlag.AlignCenter)

        self.swap_btn = SvgButton(self, icon_size=24, svg_data=arrow_right_icon)
        self.swap_btn.clicked.connect(self._swap_languages)
        self.footer_layout.addWidget(self.swap_btn)

        # 目标语种按钮：启用原生 Checkable
        self.target_lang = CoreButton(CONFIG["translator"]["default_to_lang"])
        self.target_lang.setCheckable(True)
        self.target_lang.clicked.connect(lambda: self.display_lang_list("target"))
        self.footer_layout.addWidget(self.target_lang, alignment=Qt.AlignmentFlag.AlignCenter)

        self.footer_layout.addStretch()

        default_server = CONFIG["translator"].get("default_server", "Baidu")
        self._current_server = default_server if default_server in self.SUPPORTED_SERVERS else self.SUPPORTED_SERVERS[0]

        # 翻译服务按钮：因为触发方式是右键不干预选中状态，不在配置里设 Checkable
        self.translation_server_btn = CoreButton(
            self._server_display_name(self._current_server),
            bg_color="accent",
            parent=self,
        )
        self.translation_server_btn.clicked.connect(self._start_translation)
        self.translation_server_btn.setContextMenuPolicy(Qt.ContextMenuPolicy.CustomContextMenu)
        self.translation_server_btn.customContextMenuRequested.connect(self.display_server_list)
        self.footer_layout.addWidget(self.translation_server_btn)

        self.cancel_btn = CoreButton("Cancel", parent=self)
        self.cancel_btn.setBgColor("danger")
        self.cancel_btn.hide()
        self.cancel_btn.clicked.connect(self._cancel_translation)
        self.footer_layout.addWidget(self.cancel_btn)

        self.footer_layout.setContentsMargins(0, 0, 0, 0)
        self.footer_layout.addStretch()

        layout.addWidget(self.input_text)
        layout.addSpacing(4)
        layout.addWidget(self.result_text)
        layout.addSpacing(4)
        layout.addLayout(self.footer_layout)
        layout.addWidget(self.selection_grid)

        self.one_click_hotkey = TranslationHotkey(self._on_one_click_translate, parent=self)
        self.one_click_hotkey.start()

        translation_global_signals.translate_shortcut_signal.connect(self.translate_shortcut)

    def on_show(self):
        tm = text_manager.get()
        now_time = time.perf_counter()
        elapsed = now_time - tm.selection_time
        if elapsed <= 10 and self.input_text.toPlainText() == "":
            selected_text = tm.selected_text
            if selected_text:
                self.input_text.setText(selected_text)
            elif (now_time - tm.copy_time) <= 10:
                self.input_text.setText(tm.clipboard_text)

    def _on_one_click_translate(self):
        """一键翻译：复制选中文本 → 填入输入框 → 使用默认服务翻译"""
        tm = text_manager.get()
        selected = tm.copy_selected_text()
        if not selected:
            logger.info("[TranslatorPage] 未获取到选中的文本，一键翻译已取消")
            return
        self.translate_shortcut(text=selected)

    def translate_shortcut(self, text: str):
        # 1. 切换到翻译页并展示选中文本
        page_router.immediate_switch("translator")
        self.input_text.setText(text)
        self.input_text.setFocus()

        self.origin_lang.setText(
            CONFIG["translator"]["default_from_lang"],
        )
        self.target_lang.setText(CONFIG["translator"]["default_to_lang"])

        # 2. 使用默认服务与默认语言执行翻译
        default_server = CONFIG["translator"].get("default_server", "Baidu")
        self._current_server = default_server if default_server in self.SUPPORTED_SERVERS else self.SUPPORTED_SERVERS[0]
        self.translation_server_btn.setText(self._server_display_name(self._current_server))
        self._start_translation()

    # ==================== 网格切换核心逻辑 ====================
    def display_lang_list(self, target_type="origin"):
        """一句话调度网格呈现，UI与状态交由SelectionGrid自动处理"""
        trigger_btn = self.origin_lang if target_type == "origin" else self.target_lang
        current_lang = trigger_btn.text()

        def set_language(selected_lang):
            trigger_btn.setText(selected_lang)

        extra = [(self.result_text, self.result_text.height(), 0)] if self.result_text.height() > 0 else None
        self.selection_grid.toggle(trigger_btn, self.SUPPORTED_LANGUAGES, current_lang, set_language, extra)

    def display_server_list(self):
        items = []
        self._server_display_to_id = {}
        for server_id in self.SUPPORTED_SERVERS:
            display_name = self._server_display_name(server_id)
            self._server_display_to_id[display_name] = server_id
            items.append(display_name)

        current_display = self._server_display_name(self._current_server)

        def set_server(selected_display):
            self._current_server = self._server_display_to_id[selected_display]
            self.translation_server_btn.setText(selected_display)

        extra = [(self.result_text, self.result_text.height(), 0)] if self.result_text.height() > 0 else None
        self.selection_grid.toggle(self.translation_server_btn, items, current_display, set_server, extra)

    # ==================== 杂项 ====================

    def _server_display_name(self, server_id):
        if server_id in ("AI1", "AI2"):
            name = CONFIG["translator"].get("apis", {}).get("ai", {}).get(server_id, {}).get("name")
            return name or server_id
        return server_id

    def _start_translation(self):
        if self._worker is not None:
            return

        text = self.input_text.toPlainText()
        server = self._current_server
        from_lang = self.origin_lang.text()
        to_lang = self.target_lang.text()

        logger.info(f"正在使用 [{server}] 将 '{text}' 从 {from_lang} 翻译为 {to_lang}...")

        self._translation_cancelled = False
        self._set_translating(True)

        self._worker = TranslationWorker(
            translator=self.translator,
            text=text,
            server=server,
            from_lang=from_lang,
            to_lang=to_lang,
            parent=self,
        )
        self._worker.translation_done.connect(self._on_translation_done)
        self._worker.finished.connect(self._on_worker_finished)
        self._worker.start()

    def _set_translating(self, translating: bool):
        self.translation_server_btn.setVisible(not translating)
        self.cancel_btn.setVisible(translating)

    def _cancel_translation(self):
        self._translation_cancelled = True
        worker = self._worker
        self._worker = None
        if worker is not None:
            worker.cancel()
            worker.finished.connect(worker.deleteLater)
        self._set_translating(False)

    def _on_translation_done(self, result):
        if self._translation_cancelled:
            self._translation_cancelled = False
            return

        self.result_text.setText(result)
        # 直接通知网格收起自身即可
        self.selection_grid.collapse(extra_animations=[(self.result_text, self.result_text.height(), self.RESULT_TEXT_HEIGHT)])
        self._set_translating(False)

    def _on_worker_finished(self):
        worker = self._worker
        self._worker = None
        if worker is not None:
            worker.deleteLater()

    def _swap_languages(self):
        temp = self.origin_lang.text()
        self.origin_lang.setText(self.target_lang.text())
        self.target_lang.setText(temp)

    def on_back_clicked(self):
        if self.selection_grid.is_expanded:
            self.selection_grid.collapse()
        else:
            page_router.exit_self(self.page_name)

    def clear_data(self):
        self._cancel_translation()
        self.input_text.setText("")
        self.result_text.setText("")

        self.selection_grid.collapse(extra_animations=[(self.result_text, self.result_text.height(), 0)])

File: pages/translator_page.cpp
#include "pages/translator_page.h"

#include "core/hotkey_manager.h"
#include "core/page_router.h"
#include "resources/colors.h"
#include "resources/constants.h"
#include "resources/svgs.h"
#include "utils/text_manager.h"
#include "widgets/core_button.h"
#include "widgets/svg_button.h"
#include "widgets/text_editor.h"

#include <QCursor>
#include <QEasingCurve>
#include <QFont>
#include <QGridLayout>
#include <QHBoxLayout>
#include <QParallelAnimationGroup>
#include <QPoint>
#include <QPropertyAnimation>
#include <QVariant>
#include <QWidget>
#include <QtGlobal>

#include <cmath>

// ==================== GridAnimator ====================

GridAnimator::GridAnimator(QObject* parent)
    : QObject(parent)
{
}

void GridAnimator::animateHeights(const QVector<std::tuple<QWidget*, int, int>>& animationsData,
                                  int duration, std::function<void()> onFinished)
{
    // 替换旧动画组：stop 后 deleteLater 释放 C++ 对象，
    // 否则旧组因 parent 指向常驻页面而永不回收，每次交互累积泄漏
    if (m_activeGroup) {
        m_activeGroup->stop();
        m_activeGroup->deleteLater();
        m_activeGroup = nullptr;
    }

    m_activeGroup = new QParallelAnimationGroup(this);

    for (const auto& [widget, startH, endH] : animationsData) {
        widget->setFixedHeight(startH);

        auto* anim = new QPropertyAnimation(widget, QByteArrayLiteral("maximumHeight"),
                                            m_activeGroup);
        anim->setDuration(duration);
        anim->setStartValue(startH);
        anim->setEndValue(endH);
        anim->setEasingCurve(QEasingCurve::OutQuart);

        // 同步更新 minimumHeight
        connect(anim, &QPropertyAnimation::valueChanged, widget,
                [widget](const QVariant& value) { widget->setMinimumHeight(value.toInt()); });

        // 动画结束后的边界对齐处理
        connect(anim, &QPropertyAnimation::finished, widget, [widget, endH]() {
            widget->setMinimumHeight(endH);
            widget->setMaximumHeight(endH);
        });
        m_activeGroup->addAnimation(anim);
    }

    if (onFinished)
        connect(m_activeGroup, &QParallelAnimationGroup::finished, this, onFinished);

    m_activeGroup->start();
}

// ==================== TranslationHotkey ====================

TranslationHotkey::TranslationHotkey(std::function<void()> callback, QString hotkey,
                                     QObject* parent)
    : QObject(parent)
    , m_callback(std::move(callback))
    , m_hotkey(std::move(hotkey))
{
    if (m_hotkey.isEmpty()) {
        m_hotkey = AppConfig::instance().string(QStringLiteral("translator.hotkey"),
                                                QStringLiteral("ctrl+shift+t"));
    }
}

void TranslationHotkey::start()
{
    // 注册全局热键（幂等）
    HotkeyManager& hotkeys = HotkeyManager::instance();
    hotkeys.start();
    if (hotkeys.registerHotkey(m_hotkey, [this]() {
            // HotkeyManager 回调已在 Qt 主线程分发，直接执行一键翻译流程
            if (m_callback)
                m_callback();
        })) {
        m_registered = true;
        qInfo("[TranslationHotkey] 一键翻译已启用，快捷键: %ls",
              reinterpret_cast<const wchar_t*>(m_hotkey.utf16()));
    } else {
        qWarning("[TranslationHotkey] 一键翻译快捷键 %ls 注册失败！",
                 reinterpret_cast<const wchar_t*>(m_hotkey.utf16()));
    }
}

void TranslationHotkey::stop()
{
    if (m_registered) {
        HotkeyManager::instance().unregisterHotkey(m_hotkey);
        m_registered = false;
    }
}

// ==================== TranslationWorker ====================

TranslationWorker::TranslationWorker(Translator* translator, const QString& text,
                                     const QString& server, const QString& fromLang,
                                     const QString& toLang, QObject* parent)
    : QThread(parent)
    , m_translator(translator)
    , m_text(text)
    , m_server(server)
    , m_fromLang(fromLang)
    , m_toLang(toLang)
{
}

void TranslationWorker::run()
{
    if (m_cancelled.loadRelaxed())
        return;
    const QString result = m_translator->translateText(m_text, m_fromLang, m_toLang, m_server);
    if (!m_cancelled.loadRelaxed())
        emit translationDone(result);
}

// ==================== TranslatorPage ====================

const QStringList& TranslatorPage::supportedLanguages()
{
    static const QStringList langs = {
        QStringLiteral("Auto"),    QStringLiteral("English"), QStringLiteral("Chinese"),
        QStringLiteral("Japanese"), QStringLiteral("Korean"),  QStringLiteral("French"),
        QStringLiteral("German"),  QStringLiteral("Spanish"), QStringLiteral("Russian"),
    };
    return langs;
}

const QStringList& TranslatorPage::supportedServers()
{
    static const QStringList servers = {
        QStringLiteral("Google"), QStringLiteral("DeepL"), QStringLiteral("Baidu"),
        QStringLiteral("Bing"),   QStringLiteral("AI1"),   QStringLiteral("AI2"),
    };
    return servers;
}

TranslatorPage::TranslatorPage(QWidget* parent)
    : BasePage(parent)
{
    m_translator = new Translator();
    m_animator = new GridAnimator(this);
    setTargetSize(QSize(400, 300));

    QBoxLayout* layout = setMainLayout('v');

    // 配色：激活态使用 accent 高亮，非激活态使用灰色（参考 text_editor 的暗灰配色）
    m_accentQColor = Colors::getPurestColor(Colors::getAccentColor());
    m_idleBtnBg = Colors::neutral2();

    // 文本输入框与结果框（圆角背景 + accent/灰色状态边框 + placeholder）
    m_inputText = new RoundedTextEdit(QStringLiteral("Enter or paste text here..."),
                                      Colors::neutral1(), 12, this);
    m_resultText = new RoundedTextEdit(QStringLiteral("Translation result"),
                                       Colors::neutral1(), 10, this);

    QFont font;
    font.setPointSize(12);
    m_inputText->setFont(font);
    m_resultText->setFont(font);

    m_resultText->setMinimumHeight(0);
    m_resultText->setMaximumHeight(0);

    // 通用平铺网格选择面板
    m_selectionGridWidget = new QWidget(this);
    m_gridLayout = new QGridLayout(m_selectionGridWidget);
    m_gridLayout->setContentsMargins(0, 0, 0, 10);
    m_gridLayout->setSpacing(GRID_SPACING);

    m_selectionGridWidget->setMinimumHeight(0);
    m_selectionGridWidget->setMaximumHeight(0);

    // 底部控制栏
    auto* footerLayout = new QHBoxLayout();

    footerLayout->addStretch();

    const QString defaultFrom = AppConfig::instance().string(
        QStringLiteral("translator.default_from_lang"), QStringLiteral("Auto"));
    const QString defaultTo = AppConfig::instance().string(
        QStringLiteral("translator.default_to_lang"), QStringLiteral("Chinese"));

    m_originLang = new CoreButton(defaultFrom);
    connect(m_originLang, &CoreButton::clicked, this,
            [this]() { displayLangList(QStringLiteral("origin")); });
    footerLayout->addWidget(m_originLang, 0, Qt::AlignmentFlag::AlignCenter);

    m_swapBtn = new SvgButton(this, 36, 24, Svgs::arrowRightIcon());
    connect(m_swapBtn, &SvgButton::clicked, this, &TranslatorPage::swapLanguages);
    footerLayout->addWidget(m_swapBtn);

    m_targetLang = new CoreButton(defaultTo);
    connect(m_targetLang, &CoreButton::clicked, this,
            [this]() { displayLangList(QStringLiteral("target")); });
    footerLayout->addWidget(m_targetLang, 0, Qt::AlignmentFlag::AlignCenter);

    footerLayout->addStretch();

    // 默认服务：config 指定内部标识符，按钮文本显示 config 中配置的名称
    const QString defaultServer = AppConfig::instance().string(
        QStringLiteral("translator.default_server"), QStringLiteral("Baidu"));
    m_currentServer = supportedServers().contains(defaultServer) ? defaultServer
                                                                 : supportedServers().first();

    m_translationServerBtn = new CoreButton(serverDisplayName(m_currentServer), QColor(),
                                            QColor(), 12, this);
    connect(m_translationServerBtn, &CoreButton::clicked, this,
            &TranslatorPage::startTranslation);
    m_translationServerBtn->setContextMenuPolicy(Qt::ContextMenuPolicy::CustomContextMenu);
    connect(m_translationServerBtn, &QWidget::customContextMenuRequested, this,
            [this](const QPoint&) { displayServerList(); });
    footerLayout->addWidget(m_translationServerBtn);

    // 取消按钮：与翻译按钮共存于布局，翻译时通过 hide/show 切换显示
    m_cancelBtn = new CoreButton(QStringLiteral("Cancel"), QColor(214, 69, 65), QColor(), 12,
                                 this);
    m_cancelBtn->hide();
    connect(m_cancelBtn, &CoreButton::clicked, this, &TranslatorPage::cancelTranslation);
    footerLayout->addWidget(m_cancelBtn);

    footerLayout->addStretch();

    // 布局组织
    layout->addWidget(m_inputText);
    layout->addWidget(m_resultText);
    layout->addLayout(footerLayout);
    layout->addWidget(m_selectionGridWidget);
    layout->addStretch();

    // 初始状态：网格未展开，from/to 语言按钮均置为灰色（否则默认 accent 高亮）
    setLangButtonsActive(GridMode::None);

    // 一键翻译：注册全局热键（复制选中文本 -> 填入输入框 -> 默认服务翻译）
    m_oneClickHotkey = new TranslationHotkey(
        [this]() { onOneClickTranslate(); }, QString(), this);
    m_oneClickHotkey->start();
}

void TranslatorPage::onShow()
{
    TextManager& tm = TextManager::instance();
    const double nowTime = tm.nowSec();
    const double elapsed = nowTime - tm.selectionTime;
    if (elapsed <= 10 && m_inputText->toPlainText().isEmpty()) {
        if (!tm.selectedText.isEmpty()) {
            m_inputText->setText(tm.selectedText);
        } else if (nowTime - tm.copyTime <= 10) {
            m_inputText->setText(tm.clipboardText);
        }
    }
}

void TranslatorPage::onOneClickTranslate()
{
    // 一键翻译：复制选中文本 -> 填入输入框 -> 使用默认服务翻译
    TextManager& tm = TextManager::instance();
    const QString selected = tm.copySelectedText();
    if (selected.isEmpty()) {
        qInfo("[TranslatorPage] 未获取到选中的文本，一键翻译已取消");
        return;
    }

    // 1. 切换到翻译页并展示选中文本
    PageRouter::instance().immediateSwitch(QStringLiteral("translator"));
    m_inputText->setText(selected);
    m_inputText->setFocus();

    m_originLang->setText(AppConfig::instance().string(
        QStringLiteral("translator.default_from_lang"), QStringLiteral("Auto")));
    m_targetLang->setText(AppConfig::instance().string(
        QStringLiteral("translator.default_to_lang"), QStringLiteral("Chinese")));

    // 2. 使用默认服务与默认语言执行翻译
    const QString defaultServer = AppConfig::instance().string(
        QStringLiteral("translator.default_server"), QStringLiteral("Baidu"));
    m_currentServer = supportedServers().contains(defaultServer)
                          ? defaultServer
                          : supportedServers().first();
    m_translationServerBtn->setText(serverDisplayName(m_currentServer));
    startTranslation();
}

// ==================== 抽象核心逻辑 ====================

void TranslatorPage::requestGridSwitch(GridMode mode, const QStringList& items,
                                       const QString& currentValue,
                                       const std::function<void(const QString&)>& onSelectCallback)
{
    // 网格切换控制中心：实现平滑过渡
    if (m_currentGridMode == mode) {
        collapseGrid();
        return;
    }

    const int currentHeight = prepareGridSwitch(mode, items, currentValue, onSelectCallback);
    animateGridSwitch(currentHeight, calculateGridHeight(items.size()));
}

int TranslatorPage::prepareGridSwitch(GridMode mode, const QStringList& items,
                                      const QString& currentValue,
                                      const std::function<void(const QString&)>& onSelectCallback)
{
    // 切换准备：更新网格状态、填充新按钮并固定当前高度防止跳变
    const int currentHeight = m_currentGridMode != GridMode::None
                                  ? m_selectionGridWidget->height()
                                  : 0;
    m_currentGridMode = mode;

    // 填充新按钮并强制固定当前高度防止跳变
    populateGrid(items, currentValue, onSelectCallback);
    setLangButtonsActive(mode);
    m_selectionGridWidget->setMaximumHeight(currentHeight);

    return currentHeight;
}

void TranslatorPage::animateGridSwitch(int currentHeight, int targetHeight)
{
    // 网格展开动画：网格平滑展开，同时收起结果框（若有内容）
    QVector<std::tuple<QWidget*, int, int>> animations;
    animations.append(std::make_tuple(m_selectionGridWidget, currentHeight, targetHeight));
    if (m_resultText->height() > 0)
        animations.append(std::make_tuple(m_resultText, m_resultText->height(), 0));

    m_animator->animateHeights(animations);
}

void TranslatorPage::collapseGrid(std::function<void()> onFinished)
{
    // 收起当前网格动画
    m_currentGridMode = GridMode::None;
    setLangButtonsActive(GridMode::None);
    const int currentHeight = m_selectionGridWidget->height();

    m_animator->animateHeights({std::make_tuple(m_selectionGridWidget, currentHeight, 0)}, 300,
                               std::move(onFinished));
}

void TranslatorPage::populateGrid(const QStringList& items, const QString& currentValue,
                                  const std::function<void(const QString&)>& onSelectCallback)
{
    // 清空旧按钮
    while (m_gridLayout->count()) {
        QLayoutItem* item = m_gridLayout->takeAt(0);
        if (QWidget* w = item->widget())
            w->deleteLater();
        delete item;
    }

    // 创建新按钮
    constexpr int cols = 3;
    for (int idx = 0; idx < items.size(); ++idx) {
        const QString& text = items.at(idx);
        auto* btn = new CoreButton(text);
        btn->setCursor(Qt::PointingHandCursor);
        // 仅当前选中项（激活）以 accent 高亮，其余显示灰色
        if (text != currentValue)
            btn->setBgColor(m_idleBtnBg);

        connect(btn, &CoreButton::clicked, btn, [this, text, onSelectCallback]() {
            handleGridItemClick(text, onSelectCallback);
        });

        const int row = idx / cols;
        const int col = idx % cols;
        m_gridLayout->addWidget(btn, row, col);
    }
}

int TranslatorPage::calculateGridHeight(int itemCount)
{
    constexpr int cols = 3;
    const int rows = (itemCount + cols - 1) / cols;
    return rows * GRID_ITEM_HEIGHT + (rows - 1) * GRID_SPACING;
}

void TranslatorPage::handleGridItemClick(const QString& selectedText,
                                         const std::function<void(const QString&)>& callback)
{
    // 点击网格项后的逻辑处理
    callback(selectedText);
    collapseGrid();
}

void TranslatorPage::displayLangList(const QString& targetType)
{
    // 显示语言选择网格
    const bool origin = targetType == QLatin1String("origin");
    const GridMode mode = origin ? GridMode::OriginLang : GridMode::TargetLang;
    const QString currentLang = origin ? m_originLang->text() : m_targetLang->text();

    auto setLanguage = [this, origin](const QString& selectedLang) {
        if (origin)
            m_originLang->setText(selectedLang);
        else
            m_targetLang->setText(selectedLang);
    };

    requestGridSwitch(mode, supportedLanguages(), currentLang, setLanguage);
}

QString TranslatorPage::serverDisplayName(const QString& serverId)
{
    // 服务按钮显示名：AI1/AI2 使用 config 中指定的名称，其余显示自身标识符
    if (serverId == QLatin1String("AI1") || serverId == QLatin1String("AI2")) {
        const QString name = AppConfig::instance().string(
            QStringLiteral("translator.apis.ai.") + serverId + QStringLiteral(".name"));
        return name.isEmpty() ? serverId : name;
    }
    return serverId;
}

void TranslatorPage::displayServerList()
{
    // 显示翻译服务选择网格（AI1/AI2 显示 config 指定的名称，其余显示自身名称）
    QStringList items;
    m_serverDisplayToId.clear();
    for (const QString& serverId : supportedServers()) {
        const QString displayName = serverDisplayName(serverId);
        m_serverDisplayToId.insert(displayName, serverId);
        items.append(displayName);
    }

    const QString currentDisplay = serverDisplayName(m_currentServer);

    auto setServer = [this](const QString& selectedDisplay) {
        m_currentServer = m_serverDisplayToId.value(selectedDisplay);
        m_translationServerBtn->setText(selectedDisplay);
    };

    requestGridSwitch(GridMode::Server, items, currentDisplay, setServer);
}

// ==================== 翻译流程 ====================

void TranslatorPage::startTranslation()
{
    // 后台线程执行翻译，避免阻塞 UI；翻译期间按钮替换为红色 Cancel 按钮
    if (m_worker)
        return;   // 已有翻译进行中（此时按钮已变为 Cancel，点击即取消）

    const QString text = m_inputText->toPlainText();
    const QString server = m_currentServer;
    const QString fromLang = m_originLang->text();
    const QString toLang = m_targetLang->text();

    qInfo("正在使用 [%ls] 将 '%ls' 从 %ls 翻译为 %ls...",
          reinterpret_cast<const wchar_t*>(server.utf16()),
          reinterpret_cast<const wchar_t*>(text.utf16()),
          reinterpret_cast<const wchar_t*>(fromLang.utf16()),
          reinterpret_cast<const wchar_t*>(toLang.utf16()));

    m_translationCancelled = false;
    setTranslating(true);

    m_worker = new TranslationWorker(m_translator, text, server, fromLang, toLang, this);
    connect(m_worker, &TranslationWorker::translationDone,
            this, &TranslatorPage::onTranslationDone);
    connect(m_worker, &QThread::finished, this, &TranslatorPage::onWorkerFinished);
    m_worker->start();
}

void TranslatorPage::setTranslating(bool translating)
{
    // 隐藏的组件会自动空出布局位置并触发重排，两个按钮在布局中
    // 始终占据同一槽位，因此无需移除/插入即可完成切换
    m_translationServerBtn->setVisible(!translating);
    m_cancelBtn->setVisible(translating);
}

void TranslatorPage::cancelTranslation()
{
    // 取消进行中的翻译：丢弃结果并立即恢复翻译按钮（同步请求无法中断网络传输）
    m_translationCancelled = true;
    TranslationWorker* worker = m_worker;
    m_worker = nullptr;
    if (worker) {
        worker->cancel();
        // 线程结束后自动释放，避免 QThread 对象泄漏
        connect(worker, &QThread::finished, worker, &QObject::deleteLater);
    }
    setTranslating(false);
}

void TranslatorPage::onTranslationDone(const QString& result)
{
    // 翻译完成（主线程）：显示结果并展开结果框
    if (m_translationCancelled) {
        m_translationCancelled = false;
        return;
    }

    m_resultText->setText(result);

    m_currentGridMode = GridMode::None;
    setLangButtonsActive(GridMode::None);

    const int gridStartH = m_selectionGridWidget->height();
    const int resultStartH = m_resultText->height();

    m_animator->animateHeights({
        std::make_tuple(m_selectionGridWidget, gridStartH, 0),
        std::make_tuple(m_resultText, resultStartH, RESULT_TEXT_HEIGHT),
    });
    setTranslating(false);
}

void TranslatorPage::onWorkerFinished()
{
    // 后台线程自然结束（未被取消）：释放 worker
    TranslationWorker* worker = m_worker;
    m_worker = nullptr;
    if (worker)
        worker->deleteLater();
}

void TranslatorPage::setLangButtonsActive(GridMode mode)
{
    // 仅当对应语言网格展开时，from/to 语言按钮才以 accent 高亮，否则显示灰色
    m_originLang->setBgColor(mode == GridMode::OriginLang ? m_accentQColor : m_idleBtnBg);
    m_targetLang->setBgColor(mode == GridMode::TargetLang ? m_accentQColor : m_idleBtnBg);
}

void TranslatorPage::swapLanguages()
{
    // 互换源语言与目标语言
    const QString temp = m_originLang->text();
    m_originLang->setText(m_targetLang->text());
    m_targetLang->setText(temp);
}

void TranslatorPage::onBackClicked()
{
    // 返回逻辑：如果网格开启则收起，否则退出页面
    if (m_currentGridMode != GridMode::None) {
        collapseGrid();
    } else {
        PageRouter::instance().exitSelf(m_pageName);
    }
}

void TranslatorPage::clearData()
{
    cancelTranslation();
    m_inputText->setText(QString());
    m_resultText->setText(QString());

    setLangButtonsActive(GridMode::None);

    m_animator->animateHeights({
        std::make_tuple(m_selectionGridWidget, m_selectionGridWidget->height(), 0),
        std::make_tuple(m_resultText, m_resultText->height(), 0),
    });
}
